--- boost/locale/conversion.hpp.orig	2012-11-02 13:16:47.000000000 +0100
+++ boost/locale/conversion.hpp	2012-11-02 13:16:57.000000000 +0100
@@ -143,7 +143,7 @@
             norm_nfc,   ///< Canonical decomposition followed by canonical composition
             norm_nfkd,  ///< Compatibility decomposition
             norm_nfkc,  ///< Compatibility decomposition followed by canonical composition.
-            norm_default = norm_nfc, ///< Default normalization - canonical decomposition followed by canonical composition
+            norm_default = norm_nfc ///< Default normalization - canonical decomposition followed by canonical composition
         } norm_type;
        
         ///
