Index: boost1.40-1.40.0/boost/smart_ptr/detail/spinlock_gcc_arm.hpp
===================================================================
--- boost1.40-1.40.0.orig/boost/smart_ptr/detail/spinlock_gcc_arm.hpp	2009-03-02 11:22:16.000000000 -0500
+++ boost1.40-1.40.0/boost/smart_ptr/detail/spinlock_gcc_arm.hpp	2010-03-27 15:40:14.227864125 -0400
@@ -29,11 +29,15 @@
     {
         int r;
 
+#ifdef	G_ATOMIC_ARM_USE_GCC_INTRINSICS
+		r =  __sync_lock_test_set(&v_, 1);
+#else
         __asm__ __volatile__(
             "swp %0, %1, [%2]":
             "=&r"( r ): // outputs
             "r"( 1 ), "r"( &v_ ): // inputs
             "memory", "cc" );
+#endif
 
         return r == 0;
     }
